### PR TITLE
Bug 1226901 - Add logging for corruption DBs

### PR DESCRIPTION
**Pull request by sleroux**

### CODE REVIEW

File: Storage/ThirdParty/SwiftData.swift [hunk 1]
```diff
@@ -304,6 +304,11 @@ public class SQLiteDBConnection {
         assert(cursor[0] == expected, message)
     }
 
+    private func pragma(pragma: String, factory: SDRow -> String) -> String? {
```
Comment (rnewman) [inline Storage/ThirdParty/SwiftData.swift]: Wanna rewrite the existing `pragma` to use this?

File: Storage/ThirdParty/SwiftData.swift [hunk 2]
```diff
@@ -471,6 +476,11 @@ public class SQLiteDBConnection {
             statement = nil
         }
         if let error = error {
+            // Special case: Write additional info to the database log in the case of a database corruption.
+            if error.code == 11 {
```
Comment (rnewman) [inline Storage/ThirdParty/SwiftData.swift]: `SQLITE_CORRUPT` is the constant you want.

File: Storage/ThirdParty/SwiftData.swift [hunk 3]
```diff
@@ -484,6 +494,24 @@ public class SQLiteDBConnection {
...
+        logger.error("DB file size: \(dbFileSize) bytes")
+
+        if let message = pragma("integrity_check", factory: StringFactory) {
+            logger.error("Integrity check message: \(message)")
+        }
+
+        // Write call stack
+        logger.error("Call stack: \(NSThread.callStackSymbols())")
+
+        // Write open file handlers
```
Comment (rnewman) [inline Storage/ThirdParty/SwiftData.swift]: handles

File: Storage/ThirdParty/SwiftData.swift (hunk 2 above)
Comment (rnewman) [inline Storage/ThirdParty/SwiftData.swift]: And you should add this logic to `executeChange`, too — just above.

File: Utils/Extensions/NSFileManagerExtensions.swift [hunk 4]
```diff
@@ -56,8 +56,17 @@ public extension NSFileManager {
         }
 
         // First try to get the total allocated size and in failing that, get the file allocated size
-        return itemURL.getResourceLongLongForKey(NSURLTotalFileAllocatedSizeKey)
-            ?? itemURL.getResourceLongLongForKey(NSURLFileAllocatedSizeKey)
+        if let url = url as? NSURL {
+            return allocatedFileSizeForFileAtURL(url)
```
Comment (rnewman) [inline Utils/Extensions/NSFileManagerExtensions.swift]: If you do this with an extension you can use

```
return (url as? NSURL)?.allocatedSizeForFileAtURL() ?? 0
```

File: Storage/ThirdParty/SwiftData.swift [hunk 5]
```diff
@@ -484,6 +494,24 @@ public class SQLiteDBConnection {
...
+        let openDescriptors = FSUtils.openFileDescriptors()
+        logger.error("Open file descriptors: \(openDescriptors)")
```
Comment (rnewman) [inline Storage/ThirdParty/SwiftData.swift]: After this, we should flush the log and close it.

File: Storage/ThirdParty/SwiftData.swift [hunk 6]
```diff
@@ -484,6 +494,24 @@ public class SQLiteDBConnection {
         return cursor
     }
 
+    func writeCorruptionInfoForDBNamed(dbFilename: String, toLogger logger: XCGLogger) {
+        logger.error("Corrupt DB Detected! - DB Filenamed: \(dbFilename)")
```
Comment (rnewman) [inline Storage/ThirdParty/SwiftData.swift]: I'd like to add a simple thread-safe flag to make sure we only do this once per DB. No sense spamming this full dump to disk six or seven times every time we hit it.

File: Storage/ThirdParty/SwiftData.swift (hunk 1 above)
Comment (sleroux) [inline Storage/ThirdParty/SwiftData.swift]: Sure

File: Storage/ThirdParty/SwiftData.swift (hunk 2 above)
Comment (sleroux) [inline Storage/ThirdParty/SwiftData.swift]: SQLITE_CORRUPT isn't defined here. I think it's only in sqlite3.h.

File: Storage/ThirdParty/SwiftData.swift (hunk 5 above)
Comment (sleroux) [inline Storage/ThirdParty/SwiftData.swift]: I think this writes right away - there isn't a flush or close call.

File: Utils/Extensions/NSFileManagerExtensions.swift (hunk 4 above)
Comment (sleroux) [inline Utils/Extensions/NSFileManagerExtensions.swift]: Smooth.

**Comment by sleroux** [comment]
Added PR fixes round 2.

**Comment by sleroux** [comment]
Okay - everything should be good now.

File: Storage/ThirdParty/SwiftData.swift [hunk 7]
```diff
@@ -444,6 +452,11 @@ public class SQLiteDBConnection {
             statement = nil
         }
         if let error = error {
+            // Special case: Write additional info to the database log in the case of a database corruption.
+            if error.code == Int(SQLITE_CORRUPT) {
```
Comment (rnewman) [inline Storage/ThirdParty/SwiftData.swift]: Goddamn Int32, amirite?

**Comment by rnewman** [comment]
LGTM.

**Pull request merged**